plugins/inputs/tail/tail.go

```diff
@@ -86,89 +91,88 @@ func (t *Tail) Start(acc telegraf.Accumulator) error {
 	defer t.Unlock()
 
 	t.acc = acc
-	t.tailers = make(map[string]*tail.Tail)
-
-	return t.tailNewFiles(t.FromBeginning)
-}
-
-func (t *Tail) tailNewFiles(fromBeginning bool) error {
-	var seek *tail.SeekInfo
-	if !t.Pipe && !fromBeginning {
-		seek = &tail.SeekInfo{
-			Whence: 2,
-			Offset: 0,
-		}
-	}
+	t.lines = make(chan *logline.LogLine)
+	t.files = make(map[string]bool)
 
 	var poll bool
 	if t.WatchMethod == "poll" {
 		poll = true
 	}
 
-	// Create a "tailer" for each file
+	var err error
+	t.watcher, err = watcher.NewLogWatcher(t.PollInterval.Duration, !poll, watcher.Logger(taillog.DiscardingLogger))
+	if err != nil {
+		defer close(t.lines)
+		return err
+	}
+
+	opts := []tailer.Option{tailer.Logger(taillog.DiscardingLogger)}
+	if t.FromBeginning {
+		opts = append(opts, tailer.OneShot)
+	}
+
+	t.tailer, err = tailer.New(t.lines, t.watcher, opts...)
+	if err != nil {
+		defer t.watcher.Close()
+		defer close(t.lines)
+		return err
+	}
+
+	parser, err := t.parserFunc()
+	if err != nil {
+		defer t.tailer.Close()
+		return err
+	}
+	t.wg.Add(1)
+	go t.parse(parser)
+
+	return t.tailNewFiles(t.FromBeginning)
+}
+
+func (t *Tail) tailNewFiles(fromBeginning bool) error {
+
 	for _, filepath := range t.Files {
 		g, err := globpath.Compile(filepath)
 		if err != nil {
 			t.acc.AddError(fmt.Errorf("E! Error Glob %s failed to compile, %s", filepath, err))
 		}
 		for _, file := range g.Match() {
-			if _, ok := t.tailers[file]; ok {
+			if _, ok := t.files[file]; ok {
 				// we're already tailing this file
 				continue
 			}
 
-			tailer, err := tail.TailFile(file,
-				tail.Config{
-					ReOpen:    true,
-					Follow:    true,
-					Location:  seek,
-					MustExist: true,
-					Poll:      poll,
-					Pipe:      t.Pipe,
-					Logger:    tail.DiscardingLogger,
-				})
+			err := t.tailer.TailPath(file)
 			if err != nil {
 				t.acc.AddError(err)
 				continue
 			}
 
 			log.Printf("D! [inputs.tail] tail added for file: %v", file)
 
-			parser, err := t.parserFunc()
-			if err != nil {
-				t.acc.AddError(fmt.Errorf("error creating parser: %v", err))
-			}
-
 			// create a goroutine for each "tailer"
-			t.wg.Add(1)
-			go t.receiver(parser, tailer)
-			t.tailers[tailer.Filename] = tailer
+			t.files[file] = true
```

Comment:
Currently the only way `tail` stops following a file is if that file is deleted. However, I should be removing files from that map when `Stop()` is called. I'll fix that.


### To potentially solve the potential slow resource leak problem:

The tail library I am using supports garbage collection, but is disabled by default: https://github.com/sgtsquiggs/tail/blob/master/tailer/tail.go#L385-L418

This may be useful when tailing a logrotated file:

assuming `paths = ["/var/log/apache.log"]`, once `apache.log` rotates to `apache.log.1` telegraf will continue tailing `apache.log.1` and will additionally tail `apache.log` when once Gather is next called. With GC, the tail lib will stop tailing `apache.log.1` after 24h of inactivity. It won't restart tailing `apache.log.1` unless it matches a glob or specific pathname.

This garbage collection feature could be exposed as a interval - "stop tailing files after **interval** of inactivity", which would reset when the plugin is started (upon start or after SIGHUP).